### PR TITLE
[Flang][OpenMP] Update MapInfoFinalization to use BlockArgs Interface and modify use_device_ptr/addr to be order independent

### PR DESCRIPTION
**Pull request by agozillon**
This patch primarily updates the MapInfoFinalization pass to utilise the BlockArgument interface. It also shuffles newly added arguments the MapInfoFinalization passes to the end of the BlockArg/Relevant MapInfo lists, instead of one prior to the owning descriptor type.

During this it was noted that the use_device_ptr/addr handling of target data was a little bit too order dependent so I've attempted to make it less so, as we cannot depend on argument ordering to be the same as Fortran for any future frontends.

### CODE REVIEW

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (TIFitis) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2353]: Can you please explain the change removing the if clause?

Comment (agozillon) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2353]: I currently depend on the fact that every use device ptr/addr regardless of if it has appeared in a map clause or not is inserted into the map data list here: https://github.com/llvm/llvm-project/pull/113919/files#diff-2cbb5651f4570d81d55ac4198deda0f6f7341b2503479752ef2295da3774c586R3014 to break the ordering requirements, perhaps there's a more trivial way to do so though!

I also think every clause that gets lowered should have it's information inserted, incase we depend on it at a later stage or need to check it for other information, but that might just be because it's something I've found useful!

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3010]: Nit: Since it's used for both `use_device_ptr` and `use_device_addr`, to avoid confusion.
```suggestion
          llvm::SmallVectorImpl<Value> &useDeviceVars, MapInfoData &mapInfoData,
```

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3015]: Nit: Negate condition and `continue` to reduce nesting.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:3014]:
```suggestion
          for (auto [mapClause, devicePointer, basePointer] : llvm::zip_equal(mapInfoData.MapClause, mapInfoData.DevicePointers, mapInfoData.BasePointers)) {
```

**Review by skatrak — COMMENTED** [review]
Thank you Andrew for looking into this! I just have some small nits and suggestions.

Could you add an MLIR to LLVM IR translation test that checks that the order in which a parent map info and a child appear in an e.g. `use_device_addr` clause result in equivalent IR?

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (TIFitis) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2353]: I'm a but vague on how this is necessary for ordering purposes but I can imagine how it might be needed.


Let me share my perspective:

UseDevicePtr/Addr is really just like enabling an extra flag or mapTypeMpdifier for the mapClause. It should really just be part of MapType or the map clause itself but it's not because only a subset of the directives that allow the map clause also allow the useDevPtr/Addr.

The expected/common behavior is that a variable which occurs in a UseDev clause has already been mapped in one of the MapClause. However, this isn't a rule, hence when we encounter a UseDev clause variable with no previous MapInfo for it, we create a "dummy" mapping for it. This is more in the category "no idea what the user wants here, let he/she bother".

As such, generating separate MapInfo for each and every UseDev clause doesn't really make sense. Moreover, it is only recently we switched the UseDev clauses to make use of the MapOps but that was only to leverage the vivid data type support of the mapOp and nothing else.

With this philosophy in mind I don't think we should always be adding MapInfo for each UseDev, as they are a lack of information in themselves and hold no information wortthy of preservation.

**Comment by agozillon** [comment]
Newest update attempts to address all of @skatrak's nits, and also tries to address the concerns @TIFitis mentioned in his review comments, so hopefully the changeset now seems a little more reasonable to everyone! :-)

**Review by TIFitis — APPROVED** [review]
Thanks for making the changes. I am happy with the patch 👍🏽

**Comment by agozillon** [comment]
Thank you very much @TIFitis for the review and helpful comments! :-) I'll wait for @skatrak to have a look over the nits before going ahead with landing the PR.

**Review by skatrak — APPROVED** [review]
Thank you Andrew, I see that a couple of the spots where I had comments disappeared. Since everything still works, I'm guessing these changes weren't needed, so LGTM!

**Pull request merged**
